examples/finetuning/dpo_tic_tac_toe/src/dpo_tic_tac_toe/core.py

```
@@ -0,0 +1,289 @@
+
+    This is suitable as a state-value / reward signal for RL.
+    """
+
+    assert player_val in (1, -1), "player_val must be 1 (X) or -1 (O)"
+
+    # -------- persistent cache for solved outcomes (board, side_to_move) -> {-1, 0, 1} --------
+    # This dictionary survives across function calls.
+    outcome_cache: dict[tuple[tuple[int, ...], int], float] = getattr(evaluate_board_for_player, "_outcome_cache", None)
+    if outcome_cache is None:
+        outcome_cache = {}
+        setattr(evaluate_board_for_player, "_outcome_cache", outcome_cache)
```

Ruff (0.14.8)

Do not call `setattr` with a constant attribute value. It is not any safer than normal property access.

Replace `setattr` with assignment

(B010)